### PR TITLE
Switch to C++14

### PR DESCRIPTION
**Pull request by smessmer**
This is needed for c10 as discussed in [need a better link]

### CODE REVIEW

**Review by bddppq — CHANGES_REQUESTED** [review]
Looks like there are compilation errors in Eigen:
 https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-android-ubuntu16.04-build/5282/console
and also the default compiler shipped in couple old (but common) OS doesn't support c++14:
`c++: error: unrecognized command line option ‘-std=c++14’` in
- https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-cuda9.0-cudnn7-centos7-build/5185/console
- https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-trusty-py3.5-build/8207/console

Also once c++14-only feature has been used, other projects that depend on PyTorch 1.0 (e.g https://github.com/facebookresearch/TensorComprehensions, https://github.com/pytorch/translate) will also be forced to update to c++14.

Conclusion: we are not ready.

**Comment by ezyang** [comment]
We also can't ship this until we have instructions for using a more recent version of GCC on the dev machines that most of the core devs are running on.

**Comment by ezyang** [comment]
@smessmer Can you summarize contents of the link you posted and paste them inline in the ticket?

**Comment by goldsborough** [comment]
Uh, it seems like a drastic decision to block all Ubuntu 14.04 users from compiling with their system compilers?! @soumith ?

@smessmer also beware posting things only accessible for Facebook employees, non-Facebook users should have insight into our decisions

**Comment by soumith** [comment]
i've edited the fb-internal link out, it's not useful for github consumption. Maybe paste it into a gist.github.com file or even just inline into the issue.

From the discussion last time (since a few months ago), it was seen as pretty drastic to not have 14.04 and CentOS7 customers have an out-of-the-box experience. We didn't discuss the topic since. What changed?

**Comment by xkszltl** [comment]
@soumith 
Don't worry about CentOS 7.
They have devtoolset-4/6/7 in their scl repo, which include gcc5/6/7.

**Comment by smessmer** [comment]
We're not going to do this for now, revisiting in 2019.
